les/utils/limiter.go

```diff
@@ -0,0 +1,395 @@
+		valueSelect:    NewWeightedRandomSelect(func(item interface{}) uint64 { return item.(*nodeQueue).valueWeight }),
+		nodes:          make(map[enode.ID]*nodeQueue),
+		addresses:      make(map[string]*addressGroup),
+		sumWeightLimit: sumWeightLimit,
+	}
+	l.cond = sync.NewCond(&l.lock)
+	go l.processLoop()
+	return l
+}
+
+// selectionWeights calculates the selection weights of a node for both the address and
+// the value selector. The selection weight depends on the next request weight or the
+// summed weights of recently dropped requests. relWeight is reqWeight/maxWeight.
+func selectionWeights(relWeight, value float64) (flatWeight, valueWeight uint64) {
+	var f float64
+	if relWeight <= 0.001 {
+		f = 1
+	} else {
+		f = 0.001 / relWeight
+	}
+	f *= 1000000000
+	flatWeight, valueWeight = uint64(f), uint64(f*value)
+	if flatWeight == 0 {
+		flatWeight = 1
+	}
+	return
+}
+
+// Add adds a new request to the node queue belonging to the given id. Value belongs
+// to the requesting node. A higher value gives the request a higher chance of being
+// served quickly in case of heavy load or a DDoS attack. weight is a rough estimate
+// of the serving cost of the request. A lower weight also gives the request a
+// better chance.
+func (l *Limiter) Add(id enode.ID, address string, value float64, reqWeight uint) chan chan struct{} {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	process := make(chan chan struct{}, 1)
+	if l.quit {
+		close(process)
+		return process
+	}
+	if value > l.maxValue {
+		l.maxValue = value
+	}
+	if value > 0 {
+		// normalize value to <= 1
+		value /= l.maxValue
+	}
+	if reqWeight == 0 {
+		reqWeight = 1
+	}
+	if reqWeight > l.maxWeight {
+		l.maxWeight = reqWeight
+	}
+
+	if nq, ok := l.nodes[id]; ok {
+		nq.queue = append(nq.queue, request{process, reqWeight})
+		nq.sumWeight += reqWeight
+		nq.value = value
+		if address != nq.address {
+			// known id sending request from a new address, move to different address group
+			l.removeFromGroup(nq)
+			l.addToGroup(nq, address)
+		}
+	} else {
+		nq := &nodeQueue{
+			queue:      []request{{process, reqWeight}},
+			id:         id,
+			value:      value,
+			nextWeight: reqWeight,
+			sumWeight:  reqWeight,
+			groupIndex: -1,
+		}
+		nq.flatWeight, nq.valueWeight = selectionWeights(float64(reqWeight)/float64(l.maxWeight), value)
+		if len(l.nodes) == 0 {
+			l.cond.Signal()
+		}
+		l.nodes[id] = nq
+		if nq.valueWeight != 0 {
+			l.valueSelect.Update(nq)
+		}
+		l.addToGroup(nq, address)
+	}
+	l.sumWeight += reqWeight
+	if l.sumWeight > l.sumWeightLimit {
+		l.dropRequests()
+	}
+	return process
+}
+
+// update updates the selection weights of the node queue
+func (l *Limiter) update(nq *nodeQueue) {
+	flatWeight, valueWeight := selectionWeights(float64(nq.nextWeight)/float64(l.maxWeight), nq.value)
+	ag := l.addresses[nq.address]
+	ag.update(nq, flatWeight)
+	l.addressSelect.Update(ag)
+	if valueWeight != 0 {
```

Comment:
Here have a corner case, the original valueWeight is not zero, but after that, the weight is 0.

Reply:
Right, this was a bug, I removed the `if` clause because it is unnecessary (doing a no-op `Update` is not expensive).